app/code/Magento/ConfigurableProductGraphQl/Model/Options/Collection.php

```diff
@@ -87,6 +122,22 @@ public function getAttributesByProductId(int $productId) : array
             return [];
         }
 
+        $product = $this->productRepository->getById($productId);
+        $options = $this->configurableProductHelper->getOptions(
+            $product,
+            $this->optionsMetadata->getAllowProducts($product)
+        );
+        foreach ($attributes[$productId] as $attributeIndex => $attribute) {
```

Comment:
We are adding a O(n^2) here. can't this operation be handled on the fetch() method, which is meant for gathering attribute data?